Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,839 @@
+            }
+
+            /// <summary>
+            /// Gets an <see cref="object"/> that can be used to synchronize access to the current instance.
+            /// </summary>
+            public object SyncRoot
+            {
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                get => this.buckets;
+            }
+
+            /// <summary>
+            /// Implements <see cref="StringPool.Add"/> for the current instance.
+            /// </summary>
+            /// <param name="value">The input <see cref="string"/> instance to cache.</param>
+            /// <param name="hashcode">The precomputed hashcode for <paramref name="value"/>.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public unsafe void Add(string value, int hashcode)
+            {
+                ref string target = ref TryGet(value.AsSpan(), hashcode);
+
+                if (Unsafe.AreSame(ref target, ref Unsafe.AsRef<string>(null)))
+                {
+                    Insert(value, hashcode);
+                }
+                else
+                {
+                    target = value;
+                }
+            }
+
+            /// <summary>
+            /// Implements <see cref="StringPool.GetOrAdd(string)"/> for the current instance.
+            /// </summary>
+            /// <param name="value">The input <see cref="string"/> instance with the contents to use.</param>
+            /// <param name="hashcode">The precomputed hashcode for <paramref name="value"/>.</param>
+            /// <returns>A <see cref="string"/> instance with the contents of <paramref name="value"/>.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public unsafe string GetOrAdd(string value, int hashcode)
+            {
+                ref string result = ref TryGet(value.AsSpan(), hashcode);
+
+                if (!Unsafe.AreSame(ref result, ref Unsafe.AsRef<string>(null)))
+                {
+                    return result;
+                }
+
+                Insert(value, hashcode);
+
+                return value;
+            }
+
+            /// <summary>
+            /// Implements <see cref="StringPool.GetOrAdd(ReadOnlySpan{char})"/> for the current instance.
+            /// </summary>
+            /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+            /// <param name="hashcode">The precomputed hashcode for <paramref name="span"/>.</param>
+            /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public unsafe string GetOrAdd(ReadOnlySpan<char> span, int hashcode)
+            {
+                ref string result = ref TryGet(span, hashcode);
+
+                if (!Unsafe.AreSame(ref result, ref Unsafe.AsRef<string>(null)))
```

Comment:
Can we just have an internal helper extension/polifill to make use of the easier syntax you called out `Unsafe.IsNullRef` now? 🙂

Reply:
Just to clarify on this point as I don't think I phrased that properly on Twitch - what I meant with those two new `Unsafe.IsNullRef` and `Unsafe.NullRef` APIs coming with .NET 5 is that they'll be made public at the same time .NET 5 willl, but they will not actually be exclusive to .NET 5. All the `Unsafe` APIs, while also being available "in the box" from .NET Core 3.0 and up, are actually in a separate package (`System.Runtime.CompilerServices.Unsafe`) which is pure IL and compatible with .NET Standard, in fact we're using it in this library as well, as we use all these `Unsafe` APIs on all target frameworks. So what I mean is - once November comes and the new version becomes available, we'll be able to use those new APIs everywhere in the lib, without the need for compiler directives or anything.

So I guess my question is, since the release for the 7.0 update will roughly align with .NET 5 and the internal helper would go away as soon as that is available (so almost immediately, at that point), should we include this helper anyway or at that point wouldn't it be better to just leave this like as is for the time being, possibly with an extra comment, and then just do a refactoring pass on the whole library using those two new APIs when they're available (which I planned to do anyway)?

Let me know what you think! 😊

Reply:
Ah, I see, we can wait then. We'll probably ship after .NET 5 ships, so we can pick-up the library and make the adjustment later when it's available.

Reply:
@Sergio0694 want to open an issue for tracking this update and assign it to the 7.0 milestone?

Reply:
Leaving this here for future reference:

- Issue: https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3509
- PR: https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3510